### PR TITLE
Prevent undesired `PATCH` requests when updating the state in `Worker`s or `ShootState`s

### PR DESCRIPTION
**Pull request by rfranzke**
**How to categorize this PR?**

/area scalability
/kind bug

**What this PR does / why we need it**:
This PR prevents undesired `PATCH` requests when updating the state in `Worker`s or `ShootState`s.
Only the first three commits are relevant/related - the last commit only includes a simplification in the `worker` controller package of the extensions library.

**Which issue(s) this PR fixes**:
Related to #7593

**Special notes for your reviewer**:
/cc @timuthy 

**Release note**:

```bugfix operator
An issue has been fixed which caused undesired `PATCH` requests when updating the state in the `Worker` or `ShootState` resources.
```

### CODE REVIEW

File: pkg/gardenlet/controller/shootstate/secret/reconciler.go [hunk 1]
```diff
@@ -122,8 +123,13 @@ func (r *Reconciler) reconcile(
 		Type:   "secret",
 		Data:   runtime.RawExtension{Raw: dataJSON},
 	})
-	shootState.Spec.Gardener = dataList
 
+	// If the data list does not change, do not even try to send an empty PATCH request.
+	if apiequality.Semantic.DeepEqual(shootState.Spec.Gardener, dataList) {
+		return reconcile.Result{}, nil
+	}
+
```
Comment (timuthy) [inline pkg/gardenlet/controller/shootstate/secret/reconciler.go:133]: Didn't we still observe empty patches being sent after this check? 👀

Comment (rfranzke) [reply pkg/gardenlet/controller/shootstate/secret/reconciler.go:133]: You're right, I double-checked again and adapted this check to ensure that it works as expected.

**Review by timuthy — COMMENTED** [review]
/lgtm

**Pull request merged**